### PR TITLE
Add 'subset' diff stategy option that checks the live state is a subset of the config state.

### PR DESCRIPTION
**Pull request by tomwilkie**
Part of #46, looking for feedback and discussion more than anything.

This PR adds a `--diff-strategy=subset` flag (default value is `all`, for the current behaviour).  The affect of this flag is to remove fields from the live state that are not present in the config state, reducing the diff to a subset operation.  This is how [weaveworks/kubediff](https://github.com/weaveworks/kubediff) behaves.  Note that for lists of differing lengths we preserve the live state, meaning you'll still get diff results for missing (or added) ports, containers etc.

IMO this approach turns out to be super useful - after all, if you haven't declared the desired value for a given field in you config, do you really care what its value is?  One limitation with this approach is that we won't catch labels or annotations that have been remove from config and not from live.

An alternative approach would be to "flesh out" the local config with default values for all the fields, remove the status value which will never be present locally, and then do a full diff.   This sounds super brittle to me, as you would need to keep you local kubecfg in sync with the cluster version (or make kubecfg cluster version aware).  We'd also need to maintain a black list of fields which are never going to be present locally but don't have sensible default values - for instance most of the stuff in metadata, such as `resourceVersion`, `selfLink`, `uid` etc.

WDYT?

(NB includes #53 until that is merged)

### CODE REVIEW

File: cmd/diff.go [hunk 1]
```diff
@@ -31,7 +31,10 @@ import (
 	"github.com/ksonnet/kubecfg/utils"
 )
 
+var diffStrategy string
+
 func init() {
+	diffCmd.Flags().StringVar(&diffStrategy, "diff-strategy", "all", "Diff strategy, all or subset.")
```
Comment (anguslees) [inline cmd/diff.go]: See the pattern used for other commands: Drop the `diffStrategy` global and use `flags.GetString()` below.
ie:

```go
const flagDiffStategy = "diff-strategy"

func init() {
    diffCmd.PersistentFlags().String(flagDiffStrategy, "all", "Diff strategy, all or subset.")
    // ...
}

var diffCmd = &cobra.Command{
   // ...
   RunE: func(...) {
      flags := cmd.Flags()
      diffStrategy, err := flags.GetString(flagDiffStrategy)
      if err != nil {
         return err
      }

      // ...
}
```

File: cmd/diff.go [hunk 2]
```diff
@@ -103,6 +110,41 @@ var diffCmd = &cobra.Command{
 	},
 }
 
+func removeFields(config, live interface{}) interface{} {
```
Comment (anguslees) [inline cmd/diff.go]: Can you add some tests for these?  In particular, I had to stare at `removeListFields` a bit before I was happy it was the desired behaviour (I agree it is) - some simple edge test cases might serve as documentation for the next person...

File: cmd/diff.go [hunk 3]
```diff
@@ -103,6 +110,41 @@ var diffCmd = &cobra.Command{
...
+		if !ok {
+			continue
+		}
+		result[k] = removeFields(v1, v2)
+	}
+	return result
+}
+
+func removeListFields(config, live []interface{}) []interface{} {
+	result := []interface{}{}
```
Comment (anguslees) [inline cmd/diff.go]: We know the length of the array, may as well use that fact:
```go
result := make([]interface{}, 0, len(live))
```

**Comment by tomwilkie** [comment]
Thanks for the feedback @anguslees!  Have added a bunch of tests and fixed the other comments.

**Pull request merged**